navigation/libandroid-navigation-ui/src/main/java/com/mapbox/services/android/navigation/ui/v5/instruction/InstructionView.java

```diff
@@ -0,0 +1,204 @@
+
+  private Animation slideUpTop;
+  private Animation slideDownTop;
+  private Animation rerouteSlideUpTop;
+  private Animation rerouteSlideDownTop;
+  private AnimationSet fadeInSlowOut;
+
+  private boolean showingRerouteState;
+  private boolean turnLanesHidden = false;
+
+  public InstructionView(Context context) {
+    this(context, null);
+  }
+
+  public InstructionView(Context context, @Nullable AttributeSet attrs) {
+    this(context, attrs, -1);
+  }
+
+  public InstructionView(Context context, @Nullable AttributeSet attrs, int defStyleAttr) {
+    super(context, attrs, defStyleAttr);
+    init();
+  }
+
+  @Override
+  protected void onFinishInflate() {
+    super.onFinishInflate();
+    bind();
+    initTurnLaneRecyclerView();
+    initAnimations();
+  }
+
+  public void update(RouteProgress routeProgress) {
+    if (routeProgress != null && !showingRerouteState) {
+      InstructionModel model = new InstructionModel(routeProgress);
+      maneuverImage.setImageResource(model.getManeuverImage());
+      distanceText.setText(model.getStepDistanceRemaining());
+      instructionText.setText(StringAbbreviator.abbreviate(model.getTextInstruction()));
+      addTurnLanes(model);
+    }
+  }
+
+  public void soundFabOff() {
+    soundFab.setImageResource(R.drawable.ic_sound_off);
+  }
+
+  public void soundFabOn() {
+    soundFab.setImageResource(R.drawable.ic_sound_on);
+  }
+
+  public void setSoundChipText(String text) {
+    soundChipText.setText(text);
+  }
+
+  public void showSoundChip() {
+    soundChipText.startAnimation(fadeInSlowOut);
+  }
+
+  public void show() {
+    if (this.getVisibility() == INVISIBLE) {
+      this.setVisibility(VISIBLE);
+      this.startAnimation(slideDownTop);
+    }
+  }
+
+  public void hide() {
+    if (this.getVisibility() == VISIBLE) {
+      this.startAnimation(slideUpTop);
+      this.setVisibility(INVISIBLE);
+    }
+  }
+
+  public void showRerouteState() {
+    showingRerouteState = true;
+    rerouteLayout.startAnimation(rerouteSlideDownTop);
+  }
+
+  public void hideRerouteState() {
+    showingRerouteState = false;
+    rerouteLayout.startAnimation(rerouteSlideUpTop);
+  }
+
+  private void init() {
+    inflate(getContext(), R.layout.instruction_view_layout, this);
+  }
+
+  private void bind() {
+    maneuverImage = findViewById(R.id.maneuverImageView);
+    distanceText = findViewById(R.id.distanceText);
+    instructionText = findViewById(R.id.instructionText);
+    soundChipText = findViewById(R.id.soundText);
+    soundFab = findViewById(R.id.soundFab);
+    rerouteLayout = findViewById(R.id.rerouteLayout);
+    turnLaneLayout = findViewById(R.id.turnLaneLayout);
+    rvTurnLanes = findViewById(R.id.rvTurnLanes);
+    initInstructionAutoSize();
+    addListeners();
+  }
+
+  private void addListeners() {
+    soundFab.setOnClickListener(new OnClickListener() {
+      @Override
+      public void onClick(View v) {
+        // TODO Polly stuff
```

Comment:
Remove TODO and open an issue.

Reply:
Voice instruction logic added in https://github.com/mapbox/mapbox-navigation-android/pull/194